torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp

```diff
@@ -0,0 +1,211 @@
+#include <torch/csrc/jit/passes/onnx/scalar_type_analysis.h>
+#include <torch/csrc/jit/passes/dead_code_elimination.h>
+
+namespace torch {
+namespace jit {
+
+namespace onnx {
+using namespace ::c10::onnx;
+}
+
+namespace {
+class ScalarTypeHashFunction {
+ public:
+  size_t operator()(const c10::ScalarType& type) const {
+    return static_cast<size_t>(type);
+  }
+};
+
+static const std::unordered_map<c10::ScalarType, int, ScalarTypeHashFunction> scalarTypeToONNXTypeMap = {
+    {c10::kFloat, 1},
+    {c10::kByte, 2},
+    {c10::kChar, 3},
+    {c10::kShort, 5},
+    {c10::kInt, 6},
+    {c10::kLong, 7},
+    {c10::kBool, 9},
+    {c10::kHalf, 10},
+    {c10::kDouble, 11},
+};
+
+static int64_t ScalarTypeToONNXType(const c10::ScalarType& st) {
+  int64_t onnx_type = -1;
+  if (scalarTypeToONNXTypeMap.count(st) != 0) {
+    onnx_type = scalarTypeToONNXTypeMap.find(st)->second;
+  }
+  return onnx_type;
+}
+
+static const std::unordered_set<NodeKind> arithmeticOps = {
+  onnx::Add,
+  onnx::Sub,
+  onnx::Mul,
+  onnx::Div,
+  onnx::Gemm,
+  onnx::Pow,
+  onnx::Mod,
+};
+
+static bool IsArithmeticOp(const NodeKind& nkind) {
+  return arithmeticOps.find(nkind) != arithmeticOps.end();
+}
+
+static const std::unordered_set<NodeKind> comparisonOps = {
+  onnx::Greater,
+  onnx::Less,
+  onnx::Equal,
+};
+
+static bool IsComparisonOp(const NodeKind& nkind) {
+  return comparisonOps.find(nkind) != comparisonOps.end();
+}
+
+static ProfiledTensorTypePtr CreateProfiledTensorTypeWithScalarType(
+    const ProfiledTensorTypePtr& typePtr,
+    const c10::ScalarType& scalar_type) {
+  return ProfiledTensorType::create(
+      scalar_type,
+      typePtr->device(),
+      typePtr->sizes(),
+      typePtr->strides(),
+      typePtr->requiresGrad());
+}
+
+static bool IsImplicitCastSupported(const NodeKind& nodeKind) {
+  return (arithmeticOps.find(nodeKind) != arithmeticOps.end() ||
+      comparisonOps.find(nodeKind) != comparisonOps.end());
+}
+
+static c10::optional<c10::ScalarType> PromoteScalarTypes(const std::vector<c10::ScalarType>& types) {
```

Comment:
This is a method for extending type promotion beyond pairs, to a vector of types. It may be useful beyond this pass. Should we maybe move this as a overload to c10::promoteTypes? @bddppq What do you think?

Reply:
@smessmer @zdevito do we want to promote this method to c10 level helper function?